arbos/tx_processor.go

```diff
@@ -333,6 +346,7 @@ func (p *TxProcessor) EndTxHook(gasLeft uint64, success bool) {
 	if underlyingTx != nil && underlyingTx.Type() == types.ArbitrumRetryTxType {
 		inner, _ := underlyingTx.GetInner().(*types.ArbitrumRetryTx)
 		refund := arbmath.BigMulByUint(gasPrice, gasLeft)
+		refund = arbmath.BigMin(refund, inner.MaxRefund)
```

Comment:
I think we need to subtract the used gas from the refund since it should be spent before the sender's original balance
```suggestion
		gasUsedWei := arbmath.BigMulByUint(gasPrice, gasUsed)
		maxRefund := arbmath.BigSub(inner.MaxRefund, gasUsedWei)
		maxRefund = arbmath.BigMax(maxRefund, common.Big0)
		refund = arbmath.BigMin(refund, maxRefund)
```

Reply:
or we can calculate refund by `MaxRefund - gasUsedWei` but would need change in the manual redeem code

Reply:
`gasLeft` is the unused gas, so it already factors that in

Reply:
It is possible `gasLimit * gasPrice > maxRefund` if the unsafe entrypoint is used and L2 tx gas is paid from L2 sender balancer. When gas is used, we should deduct from `maxRefund` so that `gasUsedWei + refund <= maxRefund`. 
If we assume they use the safe entrypoint, I don't think we need `maxRefund` here at all.

Reply:
resolved offline, if the sender is creating a retryable allowing X amount of gas, it is fine to spend it first because the depositor can as well just provide 0 wei for gas